pkg/resource/deploy/step_generator.go

```diff
@@ -1622,25 +1625,44 @@ func (sg *stepGenerator) generateStepsFromDiff(
 	prov plugin.Provider, goal *resource.Goal, randomSeed []byte,
 	autonaming *plugin.AutonamingOptions,
 ) ([]Step, bool, error) {
-	diff, pcs, err := sg.diff(
-		event, goal, autonaming, randomSeed,
-		urn, old, new, oldInputs, inputs, prov)
-	if pcs != nil {
-		return []Step{
-			NewDiffStep(sg.deployment, pcs, old, new, goal.IgnoreChanges),
-		}, true, nil
+	// We might be triggering a replace if we have an old state and its trigger
+	// value is different. In this case, we don't even need to bother doing a diff
+	// in this case because we know this is a replacement. Note that we treat Null
+	// as a special case for "no trigger", that is if Trigger goes from Null to
+	// anything, or from anything to Null, it does not cause a replace.
+	triggerReplace := !new.ReplacementTrigger.IsNull() && !old.ReplacementTrigger.IsNull() &&
+		!new.ReplacementTrigger.DeepEquals(old.ReplacementTrigger)
```

Comment:
How does this handle unknowns and secrets? I feel like we should strip secrets from the comparison, so that `secret(true) == true` and doesn't cause a replace, even though it "changed". Unknowns I think are fine to be treated as different. Either way we should have lifecycle tests showing these cases.